Ensure the clipboard contents are stored properly on GTK

imciner2 opened the pull request:
This PR adds the function call to tell GTK that the clipboard data can be stored. This must be done before the actual store call for the data to properly persist in some cases.

Also, this PR documents that the clipboard behavior on GTK is very dependent on the configuration of the user's machine. On my machine when testing this I noticed that the clipboard didn't persist (even with these store calls) before I installed a clipboard manager for my desktop environment. Also, in my experiments some clipboard managers (such as gpaste) will automatically treat any piece of data added to the clipboard as being persisted - making the call to `Flush` extraneous. I left a small note about that in the documentation so that developers are warned it can happen.

Closes [#10515](https://trac.wxwidgets.org/ticket/10515)

imciner2 committed:
Ensure the clipboard contents are stored properly on GTK

GTK needs to be told the data on the clipboard can be stored
before the store is attempted.

Also, document that the clipboard behavior on GTK is very
dependent on the configuration of the user's machine.

Closes #10515

vadz commented on src/gtk/clipbrd.cpp:
```diff
@@ -595,8 +597,19 @@ void wxClipboard::Clear()
+    // formed using the currently selected text.
+    if ( !m_usePrimary )
+    {
+        GtkClipboard* clipboard = gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
```
I'm a bit surprised this is not `g_clipboardAtom`. Should we use `GDK_SELECTION_CLIPBOARD` instead of it elsewhere too then?

imciner2 replied:
I don't think it matters which we use, since they both will point to the same actual `GdkAtom` I believe. We already have a mix of them though (the primary clipboard is implemented using `GDK_SELECTION_PRIMARY` everywhere). It does appear though that they have removed `gdk_atom_intern` in GTK4 (https://gitlab.gnome.org/GNOME/gtk/-/commit/54a62731916bbf724b53f0c9fc2bd1addb85802f#f49af20942c0779b40da5fc0b8af38b3ed84aacc), so the code will probably need to be updated at some point, but at that time it might be good to investigate the actual GTK clipboard API functions (https://developer.gnome.org/gtk3/stable/gtk3-Clipboards.html) instead of using the selection API functions.

vadz replied:
> both will point to the same actual `GdkAtom`

This was the part I wasn't completely sure about: `GDK_SELECTION_CLIPBOARD` is defined as `GDK_MAKE_ATOM(69)` while our `g_clipboardAtom` is `gdk_atom_intern("CLIPBOARD")`. I don't quite understand how do they manage to be the same, but if you confirm they're, this is fine. If there is any doubt, and if it also works (I have no idea how to test this), I'd rather use `g_clipboardAtom` for consistency.

vadz reviewed (approved):
Thanks for fixing this!

I'll merge it unless the comment about the clipboard atom below is worth addressing.

imciner2 committed:
Replace g_clipboardAtom with GDK_SELECTION_CLIPBOARD

GDK_SELECTION_CLIPBOARD is the prefered way of referencing
the GdkAtom for the clipboard, and using it provides a clearer
distinction between the primary clipboard and the normal clipboard.

imciner2 commented:
The `gdk_atom_intern("CLIPBOARD")` is the same as `GDK_SELECTION_CLIPBOARD`, but I agree consistency is nice. So I have now changed it so that we don't store that atom anymore and reference `GDK_SELECTION_CLIPBOARD` each time (there were only a few times we did). This is also consistent with the way the primary keyboard was implemented (it wasn't given its own global atom, it always used `GDK_SELECTION_PRIMARY`), so this should make the code easier to read and compare against the GTK documentation.

vadz commented:
Thanks! Will merge once the CI builds finish.
